### PR TITLE
Cli cache and query servers

### PR DESCRIPTION
**Pull request by savelyev-an**
1) cli cash based on TTL
2) /xxx../query_servers work differently at safe and unsafe paths, and returns only servers from projects available to user
3) remove /tenant/<uuid>/project/<uuid>/server/<uuid>

### CODE REVIEW

File: cli/internal/model/serverlist.go [hunk 1]
```diff
@@ -17,8 +18,40 @@ type ServerList struct {
...
-func SaveToFile(serverList ServerList, path string) error {
-	data, err := json.Marshal(serverList)
+type Cache struct {
+	ServerList
+	// last vault access to entity
+	TenantsTimestamps  map[iam.TenantUUID]time.Time
+	ProjectsTimestamps map[iam.ProjectUUID]time.Time
+	ServersTimestamps  map[ext.ServerUUID]time.Time
+	// ttl of entity
+	TTL time.Duration
```
Comment (shvgn) [inline cli/internal/model/serverlist.go:28]: Looks like cache-specific fields (time maps and TTL) are not used externally. Should they be exported? I suggest making them unexported.

**Review by shvgn — APPROVED** [review]
Added genearal suggestions, nothing to be blocked with

File: cli/cmd/cli/get/project.go [hunk 2]
```diff
@@ -113,12 +113,12 @@ func getProjectData(onlyCache bool, cache *model.ServerList, serverFilter model.
 		if err != nil {
 			return nil, nil, err
 		}
-		*cache = model.UpdateServerListCacheWithFreshValues(*cache, model.ServerList{
+		(*cache).Update(model.ServerList{
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply cli/cmd/cli/get/project.go:116]: Done

File: cli/internal/model/serverlist.go [hunk 3]
```diff
@@ -51,24 +85,34 @@ func TryReadCacheFromFile(filePath string) (*ServerList, error) {
...
+			ServerList: ServerList{
+				Tenants:  map[iam.TenantUUID]iam.Tenant{},
+				Projects: map[iam.ProjectUUID]iam.Project{},
+				Servers:  map[ext.ServerUUID]ext.Server{},
+			},
+			TenantsTimestamps:  map[iam.TenantUUID]time.Time{},
+			ProjectsTimestamps: map[iam.ProjectUUID]time.Time{},
+			ServersTimestamps:  map[ext.ServerUUID]time.Time{},
+			TTL:                ttl,
 		}, nil
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply cli/internal/model/serverlist.go]: fixed

File: cli/internal/model/serverlist.go [hunk 4]
```diff
@@ -17,8 +18,40 @@ type ServerList struct {
...
+			delete(c.ProjectsTimestamps, id)
+		}
+	}
+	for id, lastAccess := range c.ServersTimestamps {
+		if now.After(lastAccess.Add(c.TTL)) {
+			delete(c.Servers, id)
+			delete(c.ServersTimestamps, id)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply cli/internal/model/serverlist.go:51]: briliant, done

File: cli/internal/model/serverlist.go [hunk 5]
```diff
@@ -29,20 +62,21 @@ func SaveToFile(serverList ServerList, path string) error {
...
 	}
+	result.TTL = ttl
 	return &result, nil
 }
 
-func TryReadCacheFromFile(filePath string) (*ServerList, error) {
+func TryReadCacheFromFile(filePath string, ttl time.Duration) (*Cache, error) {
 	if _, err := os.Stat(filePath); os.IsNotExist(err) {
 		dirPath := path.Dir(filePath)
 		if _, err := os.Stat(dirPath); os.IsNotExist(err) {
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply cli/internal/model/serverlist.go:82]: fix

File: cli/internal/model/serverlist.go (hunk 1 above)
Comment (savelyev-an) [reply cli/internal/model/serverlist.go:28]: done

**Pull request merged**